### PR TITLE
Add place.ty() and Ty build from a kind to smir

### PR DESCRIPTION
**Pull request by celinval**
Add a method to retrieve the type of a place and a few utility functions needed to build the projection type. I decided to return a result to avoid panicking if the user passes invalid inputs, such as wrong list of locals.

r? @spastorino

### CODE REVIEW

File: compiler/rustc_smir/src/rustc_internal/internal.rs [hunk 1]
```diff
@@ -183,6 +259,17 @@ impl<'tcx> RustcInternal<'tcx> for ClosureKind {
...
 }
 
+impl<'tcx> RustcInternal<'tcx> for AdtDef {
+    type T = rustc_ty::AdtDef<'tcx>;
+    fn internal(&self, tables: &mut Tables<'tcx>) -> Self::T {
+        let ty = tables.tcx.type_of(self.0.internal(&mut *tables)).instantiate_identity().kind();
+        let rustc_ty::TyKind::Adt(def, _) = ty else {
+            panic!("Expected an ADT definition, but found: {ty:?}")
+        };
+        *def
```
Comment (compiler-errors) [inline compiler/rustc_smir/src/rustc_internal/internal.rs:269]: Please use the `TyCtxt::adt_def` query, roundtripping through `TyCtxt::type_of` seems a bit... roundabout 😄

**Review by compiler-errors — CHANGES_REQUESTED** [review]

**Comment by ouz-a** [comment]
Thanks this is great, left few small nits.

**Review by ouz-a — DISMISSED** [review]
few nits

File: compiler/rustc_smir/src/rustc_internal/internal.rs (hunk 1 above)
Comment (celinval) [reply compiler/rustc_smir/src/rustc_internal/internal.rs:269]: Totally, thanks for the suggestion.

File: compiler/rustc_smir/src/rustc_smir/mod.rs [hunk 2]
```diff
@@ -298,6 +299,37 @@ impl<'tcx> Context for TablesWrapper<'tcx> {
...
+            .ok_or_else(|| Error::new(format!("Const `{cnst:?}` cannot be encoded as u64")))
+    }
+
+    fn usize_to_const(&self, val: u64) -> Result<Const, Error> {
+        let mut tables = self.0.borrow_mut();
+        let ty = tables.tcx.types.usize;
+        let size = tables.tcx.layout_of(ParamEnv::empty().and(ty)).unwrap().size;
+
+        let scalar = ScalarInt::try_from_uint(val, size).ok_or_else(|| {
+            Error::new(format!("Value overflow: cannot covert `{val}` to usize."))
```
[the remark this answers is not in the document]
Comment (celinval) [reply compiler/rustc_smir/src/rustc_smir/mod.rs:322]: I am pretty sure I made that typo mistake before. :facepalm:

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
Needs a rebase, r=me after that, unless you want further review from ouz-a

**Comment by celinval** [comment]
@ouz-a, I applied your suggestions. Is there any further changes you would like me to make?

**Comment by ouz-a** [comment]
> @ouz-a, I applied your suggestions. Is there any further changes you would like me to make?

lgtm

**Pull request merged**